clarity/src/vm/analysis/type_checker/tests/mod.rs

```diff
@@ -3039,3 +3053,131 @@ fn test_principal_construct() {
+            )
+        )
+        (define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (contract-call? f foo)
+        )
+        (define-public (call-foo-foobar (f <trait-foobar>))
+            (begin
+                (call-foo f)
+            )
+        )",
+    ];
+
+    let bad = ["(define-trait trait-bar
+            (
+                (bar () (response uint uint))
+            )
+        )
+        (define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (contract-call? f foo)
+        )
+        (define-public (call-foo-foobar (f <trait-bar>))
+            (begin
+                (call-foo f)
+            )
+        )"];
+
+    let contract_identifier = QualifiedContractIdentifier::transient();
+    let bad_expected = [CheckErrors::IncompatibleTrait(
+        TraitIdentifier {
+            name: ClarityName::from("trait-foo"),
+            contract_identifier: contract_identifier.clone(),
+        },
+        TraitIdentifier {
+            name: ClarityName::from("trait-bar"),
+            contract_identifier: contract_identifier.clone(),
+        },
+    )];
+
+    for good_test in good.iter() {
+        assert!(mem_type_check(&good_test).is_ok());
+    }
+
+    for (bad_test, expected) in bad.iter().zip(bad_expected.iter()) {
+        assert_eq!(expected, &mem_type_check(&bad_test).unwrap_err().err);
+    }
+}
+
+#[test]
+fn test_wrapped_trait() {
+    let good = [
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-public (call-foo-if (opt (optional <trait-foo>)))
+            (match opt
+                f (contract-call? f foo)
+                (ok u1)
+            )
+        )",
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (call-foo (f <trait-foo>))
+            (unwrap! (contract-call? f foo) u2)
+        )
+        (define-public (call-foo-list (l (list 5 <trait-foo>)))
+            (ok (map call-foo l))
+        )",
+        "(define-trait trait-foo ((foo () (response uint uint))))
+        (define-private (return-f (f <trait-foo>))
+            (if true (ok f) (err u1))
+        )
+        (define-public (call-foo (f <trait-foo>))
+            (match (return-f f)
+                f-prime (contract-call? f-prime foo)
+                e (err u1)
+            )
+        )",
+    ];
+
+    for good_test in good.iter() {
+        assert!(mem_type_check(&good_test).is_ok());
+    }
+}
+
+#[test]
+fn test_let_bind_trait() {
+    let good = ["(define-trait trait-foo ((foo () (response uint uint))))
+        (define-public (call-foo (f <trait-foo>))
+            (let ((g f))
+                (contract-call? g foo)
+            )
+        )"];
+
+    for good_test in good.iter() {
+        assert!(mem_type_check(&good_test).is_ok());
+    }
+}
+
+#[test]
+fn test_trait_same_contract() {
```

Comment:
I'm not sure what's being tested here that isn't tested in `test_let_bind_trait()`?  Is it that we're testing that `trigger` will type-check in addition to the `contract-call?` to `g` bound to `f`?

Reply:
I guess this test should have gone before the others that use the same functionality (I believe it was written first), but the idea was just to make sure that a trait defined in the same contract can be used both in the function parameter type checking, and in a contract call. It should be the simplest check for basic trait functionality, if it fails, other trait tests should be expected to fail as well.